gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java

```diff
@@ -0,0 +1,108 @@
+
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal;
+import org.apache.tinkerpop.gremlin.process.traversal.Traverser;
+import org.apache.tinkerpop.gremlin.process.traversal.step.TraversalParent;
+import org.apache.tinkerpop.gremlin.process.traversal.traverser.TraverserRequirement;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Reference implementation for substring step, a mid-traversal step which returns a substring of the incoming string
+ * traverser with a 0-based start index (inclusive) and length specified. If the start index is negative then it will
+ * begin at the specified index counted from the end of the string, or 0 if exceeding the string length.
+ * Length is optional, if it is not specific or if it exceeds the length of the string then all remaining characters will
+ * be returned. Length <= 0 will return the empty string. Null values are not processed and remain as null when returned.
+ * If the incoming traverser is a non-String value then an {@code IllegalArgumentException} will be thrown.
+ *
+ * @author David Bechberger (http://bechberger.com)
+ * @author Yang Xia (http://github.com/xiazcy)
+ */
+public final class SubstringStep<S> extends ScalarMapStep<S, String> {
+
+    private final Long start;
+    private final Long length;
+
+    public SubstringStep(final Traversal.Admin traversal, final Long startIndex, final Long length) {
+        super(traversal);
+        this.start = startIndex;
+        this.length = length;
+    }
+
+    public SubstringStep(final Traversal.Admin traversal, final Long startIndex) {
+        this(traversal, startIndex, null);
+    }
+
+    @Override
+    protected String map(final Traverser.Admin<S> traverser) {
+        final S item = traverser.get();
+        // throws when incoming traverser isn't a string
+        if (null != item && !(item instanceof String)) {
+            throw new IllegalArgumentException(
+                    String.format("The substring() step can only take string as argument, encountered %s", item.getClass()));
+        }
+
+        // to preserve null items
+        final String strItem = (String) item;
+
+        if (null == strItem)
+            return null;
+
+        final int newStart = processStartIndex(strItem.length());
+        if (null == this.length)
+            return strItem.substring(newStart);
+
+        // length < 0 will return the empty string.
+        if (this.length <= 0)
+            return "";
+
+        // if length specified exceeds the string length it is assumed to be equal to the length, which returns all
+        // remaining characters in the string.
+        return strItem.substring(newStart,
+                Math.toIntExact(Math.min(this.length + newStart, strItem.length())));
+    }
+
+    @Override
+    public Set<TraverserRequirement> getRequirements() {
+        return Collections.singleton(TraverserRequirement.OBJECT);
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
may be better to start using recommended java way for calculation hashcode? 
something like `result = 31 * result + ...`

Reply:
Updated